apollo-router/src/axum_factory/axum_http_server_factory.rs

```diff
@@ -304,6 +327,51 @@ where
     Ok(ListenAddrAndRouter(listener, route))
 }
 
+async fn entitlement_handler<B>(
+    State((entitlement, start, delta)): State<(EntitlementState, Instant, Arc<AtomicU64>)>,
+    request: Request<B>,
+    next: Next<B>,
+) -> Response {
+    if matches!(
+        entitlement,
+        EntitlementState::EntitledHalt | EntitlementState::EntitledWarn
+    ) {
+        ::tracing::error!(
+           monotonic_counter.apollo_router_http_requests_total = 1u64,
+           status = %500u16,
+           error = "Apollo entitlement expired: {ENTITLEMENT_EXPIRED_URL}",
+        );
+
+        // This will rate limit logs about entitlement to 1 a second.
+        // The way it works is storing the delta in seconds from a starting instant.
+        // If the delta is over one second from the last time we logged then try and do a compare_exchange and if successfull log.
+        // If not successful some other thread will have logged.
+        let last_elapsed_seconds = delta.load(Ordering::Relaxed);
+        let elapsed_seconds = start.elapsed().as_secs();
+        if elapsed_seconds > last_elapsed_seconds
+            && delta
+                .compare_exchange(
+                    last_elapsed_seconds,
+                    elapsed_seconds,
+                    Ordering::Relaxed,
+                    Ordering::Relaxed,
```

Comment:
It's ok to use `Relaxed` if you are just incrementing a counter and it doesn't matter much if instructions are re-ordered. However, here you are trying to do a CAS and it may matter.

Because I'm always nervous about this kind of thing, I tend to make Ordering `SeqCst` unless I really do know that it's safe to relax it.

Note: This kind of problem won't manifest on x86 due to processor limitations, but it's very likely to trigger on M1 if there's a race.

Reply:
[21e7e93](https://github.com/apollographql/router/pull/2596/commits/21e7e93fa0bbc73ec14ab05385155caf8bcd6f34)